internal/pkg/provision/access/adapter.go

```diff
@@ -0,0 +1,104 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package access
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+
+	"github.com/talos-systems/talos/cmd/osctl/pkg/client"
+	"github.com/talos-systems/talos/internal/pkg/provision"
+	"github.com/talos-systems/talos/pkg/constants"
+)
+
+// NewAdapter returns ClusterAccess object from Cluster.
+func NewAdapter(cluster provision.Cluster) provision.ClusterAccess {
+	return &adapter{
+		Cluster: cluster,
+		clients: make(map[string]*client.Client),
+	}
+}
+
+type adapter struct {
+	provision.Cluster
+
+	clients   map[string]*client.Client
+	clientset *kubernetes.Clientset
+}
+
+func (a *adapter) Client(endpoints ...string) (*client.Client, error) {
+	key := strings.Join(endpoints, ",")
+
+	if cli := a.clients[key]; cli != nil {
+		return cli, nil
+	}
+
+	configContext, creds, err := client.NewClientContextAndCredentialsFromParsedConfig(a.TalosConfig(), "")
+	if err != nil {
+		return nil, err
+	}
+
+	if len(endpoints) == 0 {
+		endpoints = configContext.Endpoints
+	}
+
+	client, err := client.NewClient(creds, endpoints, constants.ApidPort)
+	if err == nil {
+		a.clients[key] = client
+	}
+
+	return client, err
+}
+
+func (a *adapter) K8sClient(ctx context.Context) (*kubernetes.Clientset, error) {
```

Comment:
We have a kubernetes package we maintain, should we move some of this logic in there? I could go either way, just pointing out the package (`pkg/kubernetes`).

Reply:
I missed that, looks like I can move lots of stuff there

Reply:
Looked into this more, and I couldn't find good way here. Either move all the assertions to `pkg/kubernetes` which looked strange as they depend on `ClusterAccess` object (we could first derive k8s client, and then call `pkg/kubernetes`), and these assertions seems to be specific to cluster readiness, not really generic.

Another option is to extract common code of assertions, but they seem to be trivial (getting list of nodes, pods), which is just 2-3 lines of code.

Reply:
SGTM, let's leave as is then.